diet_optimization/diet_optimization_lp_py.ipynb

```diff
@@ -0,0 +1,670 @@
+            "ice cream                   330.0                    8.0               10.0   \n",
+            "\n",
+            "           sodium (per serving)  \n",
+            "hamburger                 730.0  \n",
+            "chicken                  1190.0  \n",
+            "hot dog                  1800.0  \n",
+            "fries                     270.0  \n",
+            "macaroni                  930.0  \n",
+            "pizza                     820.0  \n",
+            "salad                    1230.0  \n",
+            "milk                      125.0  \n",
+            "ice cream                 180.0  \n"
+          ]
+        }
+      ],
+      "source": [
+        "# Create a DataFrame for better visualization\n",
+        "nutrition_df = pd.DataFrame(nutrition_data, index=categories.keys()).T\n",
+        "nutrition_df.columns = [f\"{cat} (per serving)\" for cat in categories.keys()]\n",
+        "print(\"Nutritional Values per Serving:\")\n",
+        "print(nutrition_df)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "## Problem Formulation\n",
+        "\n",
+        "Now we'll create the optimization problem using the cuOpt Python API as LP. The problem has:\n",
+        "- **Variables**: Amount of each food to buy (continuous, non-negative)\n",
+        "- **Objective**: Minimize total cost\n",
+        "- **Constraints**: Meet nutritional requirements (minimum and maximum bounds)\n"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 22,
+      "metadata": {},
+      "outputs": [
+        {
+          "name": "stdout",
+          "output_type": "stream",
+          "text": [
+            "Created 9 decision variables for foods\n",
+            "Variables: ['hamburger', 'chicken', 'hot dog', 'fries', 'macaroni', 'pizza', 'salad', 'milk', 'ice cream']\n"
+          ]
+        }
+      ],
+      "source": [
+        "# Create the optimization problem\n",
+        "problem = Problem(\"diet_optimization\")\n",
+        "\n",
+        "# Add decision variables for each food (amount to buy)\n",
+        "buy_vars = {}\n",
+        "for food_name in food_costs:\n",
+        "    var = problem.addVariable(name=f\"{food_name}\", vtype=VType.CONTINUOUS, lb=0.0, ub=float('inf'))\n",
+        "    buy_vars[food_name] = var\n",
+        "\n",
+        "print(f\"Created {len(buy_vars)} decision variables for foods\")\n",
+        "print(f\"Variables: {[var.getVariableName() for var in buy_vars.values()]}\")\n"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 23,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "# Set objective function: minimize total cost\n",
+        "obj_vars = []\n",
+        "obj_coeffs = []\n",
+        "\n",
+        "for var in buy_vars.values():\n",
+        "    if food_costs[var.getVariableName()] != 0:  # Only include non-zero coefficients\n",
```

Comment:
I think it's counterintuitive to add LinearExpressions like this. Why not generate the expression in the loop itself
expr = expr + coeff * var

We could add it for the following addConstraint calls as well

Reply:
actually, expr += coeff * var (it is internally more efficient than all the copying in expr = expr + coeff * var)